### PR TITLE
Tokenizer comments

### PR DESCRIPTION
**Pull request by smola**
## Description

Comment style is an important syntactic feature of programming languages. This commit changes the tokenizer to output a special token for each comment. A comment such as `/* Comment */` will generate the token `COMMENT/*`.

This already improves detection of some sames. For example, if Mathematica/PacletInfo.m is excluded from training, it will be classified as Mercury instead of Mathematica (on test_classify_ambiguous_languages), but after this commit, it will be classified correctly, because the `COMMENT(*` tokens provide enough information to the classifier to make the correct decision.

Tests are adjusted to:

- Ignore classifier tests for extensions where we have a catch-all rule. This includes .sql, where the classifier output is mostly useless.
- Allow `samples/C++/rpc.h` to be classified as C. This file is fully valid C without any single C++-specific syntax or convention.

## Checklist:

_Checklist does not apply._

### CODE REVIEW

**Comment by smola** [comment]
This PR is currently stacked on top of #5006 and #5060. Cross validation errors are down to 36 (from 58 in master and 45 in #5060).

<details><summary>diff report</summary>

```diff
--- tokenizer-symbols3.list	2020-10-22 21:58:34.013900257 +0200
+++ tokenizer-comments2.list	2020-10-23 00:31:04.483269226 +0200
@@ -23 +23 @@
-Assembly/macros.inc BAD(NASL)
+Assembly/macros.inc GOOD
@@ -39 +39 @@
-C#/AssemblyInfo.cs BAD(Smalltalk)
+C#/AssemblyInfo.cs GOOD
@@ -75 +75 @@
-C++/libcanister.h BAD(C)
+C++/libcanister.h GOOD
@@ -114 +114 @@
-C++/rpc.h GOOD
+C++/rpc.h BAD(C)
@@ -197 +197 @@
-GLSL/recurse1.fs BAD(Filterscript)
+GLSL/recurse1.fs GOOD
@@ -256 +256 @@
-INI/ms.properties BAD(Java Properties)
+INI/ms.properties GOOD
@@ -295 +295 @@
-Mathematica/Init.m BAD(Objective-C)
+Mathematica/Init.m GOOD
@@ -429 +429 @@
-Objective-C/Siesta.h BAD(C++)
+Objective-C/Siesta.h BAD(C)
@@ -505 +505 @@
-Proguard/proguard-rules2.pro BAD(IDL)
+Proguard/proguard-rules2.pro GOOD
@@ -558 +558 @@
-Rebol/booters.r BAD(R)
+Rebol/booters.r GOOD
@@ -624,2 +624,2 @@
-Text/min-help.ncl BAD(NCL)
-Text/rmMonAnnCycLLT-help.ncl BAD(NCL)
+Text/min-help.ncl GOOD
+Text/rmMonAnnCycLLT-help.ncl GOOD
```

</details>

**Comment by smola** [comment]
After these changes, files discussed in #5015 are correctly classified as C without any further sample.

**Comment by Alhadis** [comment]
Good job! Not familiar with this part of the codebase, but here's how Roff comments are matched:

~~~regexp
(?:^|\n)[.'][ \t]*\\" (?# Line comment)
|
(?:^|\n)\.ig\n        (?# Block comment, begin)
(?:^|\n)\.\.\n        (?# Block comment, end)
~~~

(These expressions are simplified, but they'll match the most obvious-looking cases)

**Review by JT0407 — APPROVED** [review]

**Comment by smola** [comment]
@Alhadis Added tokenization for Roff comments. You may want to check if I got it right in the new test cases.

**Comment by Alhadis** [comment]
👍 Yup, nailed it.

**Comment by smola** [comment]
#5060 does not pass tests standalone, so I could create a new PR with both changes from  #5060 and  #5061. In principle they are logically different changes, but in terms of accuracy improvement and tests, they work best together.

**Review by lildude — CHANGES_REQUESTED** [review]
>   * Allow `samples/C++/rpc.h` to be classified as C. This file is fully valid C without any single C++-specific syntax or convention.

If there is nothing C++-specific in this file, I think we should move it to C and adjust the tests accordingly. Given the ongoing problems with the header files in the C-family of languages, it might be prudent to start erring on the side of C by default unless otherwise told or detected. This could be enhanced further if @Alhadis's idea at https://github.com/github/linguist/issues/1626#issuecomment-755243516 is feasible (I don't know enough about C languages to be 💯 sure).

Other than that, a quick look at the code doesn't show any glaring issues, but updating the `samples.json` does introduce some new warnings that need addressing:

```console
$ rm -rf tmp/x86_64-darwin19/
$ bundle exec rake samples
mkdir -p tmp/x86_64-darwin19/linguist/2.7.1
cd tmp/x86_64-darwin19/linguist/2.7.1
/Users/lildude/.rbenv/versions/2.7.1/bin/ruby -I. ../../../../ext/linguist/extconf.rb
creating Makefile
cd -
cd tmp/x86_64-darwin19/linguist/2.7.1
/usr/bin/make
compiling ../../../../ext/linguist/lex.linguist_yy.c
lex.linguist_yy.c:1264:13: warning: misleading indentation; statement is not part of the previous 'if' [-Wmisleading-indentation]
            if ( ! yyg->yy_state_buf )
            ^
lex.linguist_yy.c:1262:9: note: previous statement is here
        if ( ! yyg->yy_state_buf )
        ^
1 warning generated.
compiling ../../../../ext/linguist/linguist.c
linking shared-object linguist/linguist.bundle
cd -
mkdir -p tmp/x86_64-darwin19/stage/lib/linguist
[...]
```

**Comment by smola** [comment]
@smola What's your compiler? Do you use custom CFLAGS? As far as I can tell, these misleading indentations are in flex-generated control where we have little control.

**Comment by lildude** [comment]
Xcode 12. I don't believe I've deliberately set any custom CFLAGs in a looong time. I've uploaded a copy of the generated Makefile to https://gist.github.com/lildude/eafe81e2339b6c5e1009b56e157fa72d if you want to take a look.

**Comment by lildude** [comment]
@smola I've just noticed this doesn't happen with your [staging branch](https://github.com/smola/linguist/tree/staging) so I suspect something across the three PRs "fixes" this.

File: test/test_classifier.rb [hunk 1]
```diff
@@ -54,17 +58,23 @@ def test_classify_ambiguous_languages
...
       languages = Language.find_by_extension(sample[:path]).map(&:name)
       next if languages.length <= 1
 
-      next if skip_extensions.include? sample[:extname]
-
       results = Classifier.classify(Samples.cache, File.read(sample[:path]), languages)
-      assert_equal language.name, results.first[0], "#{sample[:path]}\n#{results.inspect}"
+
+      if allowed_failures.has_key? sample[:path]
+        # Failures are reasonable when a file is fully valid in more than one language.
```
Comment (lildude) [inline test/test_classifier.rb:73]: Can this comment be moved up to where `allowed_failures` is defined

File: test/test_file_blob.rb [hunk 2]
```diff
@@ -643,11 +643,20 @@ def test_documentation
...
+      "#{samples_path}/C++/rpc.h" => ["C", "C++"],
+    }
     Samples.each do |sample|
       blob = sample_blob(sample[:path])
       assert blob.language, "No language for #{sample[:path]}"
       fs_name = blob.language.fs_name ? blob.language.fs_name : blob.language.name
-      assert_equal sample[:language], fs_name, blob.name
+
+      if allowed_failures.has_key? sample[:path]
+        # Failures are reasonable when a file is fully valid in more than one language.
```
Comment (lildude) [inline test/test_file_blob.rb:655]: Please move this comment up too.

**Review by lildude — COMMENTED** [review]
Two little nits, but otherwise looking good 😍

**Review by lildude — APPROVED** [review]
LGTM.

**Pull request merged**